developer-workflow/c-api.rst

```diff
@@ -134,6 +146,103 @@ When moving existing tests, feel free to replace ``TestError`` with
+  This happens automatically, based on the name
+  (via :cpy-file:`Doc/tools/extensions/c_annotations.py`).
+
+Despite being “unstable”, there are rules to make sure third-party code can
+use this API reliably:
+
+* Changes and removals can be done in minor releases (3.x.0, including
+  Alphas and Betas for 3.x.0).
+* Adding a new unstable API *for an existing feature* is allowed even after
+  Beta feature freeze, up until the first Release Candidate.
+  Consensus on the `Core Development Discourse <https://discuss.python.org/c/core-dev/23>`_
+  is needed in the Beta period.
+* Backwards-incompatible changes should make existing C callers fail to compile.
+  For example, arguments should be added/removed, or a function should be
+  renamed.
+* When moving an API into or out of the Unstable tier, the old name
+  should continue to be available (but deprecated) until an incompatible
+  change is made. In other words, while we're allowed to break calling code,
+  we shouldn't break it *unnecessarily*.
+
+
+Moving an API from the public tier to Unstable
+----------------------------------------------
+
+* Expose the API under its new name, with the ``PyUnstable_`` prefix.
+* Make the old name an alias (e.g. a ``static inline`` function calling the
+  new function).
+* Deprecate the old name, typically using :c:macro:`Py_DEPRECATED`.
+* Announce the change in the "What's New".
+
+The old name should continue to be available until an incompatible change is
+made. Per Python’s backwards compatibility policy (:pep:`387`),
+this deprecation needs to last at least two releases
+(modulo Steering Council exceptions).
+
+The rules are relaxed for APIs that were introduced in Python versions
+before 3.12, when the official Unstable tier was added.
+You can make an incompatible change (and remove the old name)
+as if the function was already part of the Unstable tier
+for APIs introduced before Python 3.12 that are either:
+
+* Documented to be less stable than default.
+* Named with a leading underscore.
+
+Moving an API from the private tier to unstable
+-----------------------------------------------
+
+* Expose the API under its new name, with the ``PyUnstable_`` prefix.
+* If the old name is documented, or widely used externally,
+  make it an alias and deprecate it (typically with :c:macro:`Py_DEPRECATED`).
+  It should continue to be available until an incompatible change is made,
+  as if it was previously public.
+
+  This applies even to underscored names. Python wasn't always strict with
+  the leading underscore.
+* Announce the change in What's New.
+
+Moving an API from unstable to public
+-------------------------------------
+
+* Expose the API under its new name, without the ``PyUnstable_`` prefix.
+* Make the old ``PyUnstable_*`` name be an alias (e.g. a ``static inline``
+  function calling the new function).
+* Announce the change in What's New.
+
+The old name should continue to be available until an incompatible change is
+made.
```

Comment:
Such incompatible changes after a tier promotion should be rare though, since the unstable API entry is an alias for a *public* API at that point. Suggested wording tweak:

```
The previously unstable name should continue to be made available as long as the new public
name remains available. If the public name is ever subsequently deprecated, then its unstable
alias will be deprecated at the same time.

While the unstable name could theoretically be removed or deprecated immediately without
violating the expectations set for the unstable API tier, there's also no need to break working
code just because a previously unstable API is now considered ready for a wider audience.
```

Reply:
I of course have no subject matter expert opinion, but I've [made this into an actual suggestion, on the correct lines](https://github.com/python/devguide/pull/1060#discussion_r1133040798) along with some additional copyediting and an alternate suggestion to still include the existing mention if desired.